File: app/src/main/java/io/github/lee0701/heonot/InputMethodLoader.java
package io.github.lee0701.heonot;

import org.json.JSONException;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import io.github.lee0701.heonot.inputmethod.InputMethod;

public class InputMethodLoader {

	public static InputMethod loadMethod(File methodFile) {
		try(FileInputStream fis = new FileInputStream(methodFile)) {
			byte[] bytes = new byte[fis.available()];
			fis.read(bytes);
			InputMethod method = InputMethod.loadJSON(new String(bytes));
			return method;
		} catch(IOException | JSONException e) {
			e.printStackTrace();
			return null;
		}
	}

	public static void storeMethod(File methodFile, InputMethod method) {
		try(FileOutputStream fos = new FileOutputStream(methodFile)) {
			fos.write(method.toJSON(-1).getBytes());
		} catch(IOException | JSONException e) {
			e.printStackTrace();
		}
	}

	public static List<InputMethod> loadMethods(File methodsDir) {
		List<InputMethod> inputMethods = new ArrayList<>();
		for(File file : methodsDir.listFiles()) {
			if(file.getName().endsWith(".json")) {
				String fileName = file.getName().replace(".json", "");
				try {
					int index = Integer.parseInt(fileName);
					InputMethod method = loadMethod(file);
					if(method != null) inputMethods.add(index, method);
				} catch(NumberFormatException e) {
					continue;
				}
			}
		}
		return inputMethods;
	}

	public static void storeMethods(File methodsDir, List<InputMethod> inputMethods) {
		for(File file : methodsDir.listFiles()) {
			file.delete();
		}
		for(int i = 0 ; i < inputMethods.size() ; i++) {
			InputMethod method = inputMethods.get(i);
			File file = new File(methodsDir, i + ".json");
			storeMethod(file, method);
		}
	}

}

File: heonot-core/src/main/java/io/github/lee0701/heonot/InputMethodLoader.kt
package io.github.lee0701.heonot

import org.json.JSONException

import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException

import io.github.lee0701.heonot.inputmethod.InputMethod

object InputMethodLoader {

	fun loadMethod(methodFile: File): InputMethod? {
		try {
			FileInputStream(methodFile).use { fis ->
				val bytes = ByteArray(fis.available())
				fis.read(bytes)
				return InputMethod.loadJSON(String(bytes))
			}
		} catch (e: IOException) {
			e.printStackTrace()
			return null
		} catch (e: JSONException) {
			e.printStackTrace()
			return null
		}

	}

	fun storeMethod(methodFile: File, method: InputMethod) {
		try {
			FileOutputStream(methodFile).use { fos -> fos.write(method.toJSON(-1).toByteArray()) }
		} catch (e: IOException) {
			e.printStackTrace()
		} catch (e: JSONException) {
			e.printStackTrace()
		}

	}

	fun loadMethods(methodsDir: File): MutableList<InputMethod> {
		val inputMethods = mutableListOf<InputMethod>()
		for (file in methodsDir.listFiles()!!) {
			if (file.name.endsWith(".json")) {
				val fileName = file.name.replace(".json", "")
				try {
					val index = Integer.parseInt(fileName)
					val method = loadMethod(file)
					if (method != null) inputMethods.add(index, method)
				} catch (e: NumberFormatException) {
					continue
				}

			}
		}
		return inputMethods
	}

	fun storeMethods(methodsDir: File, inputMethods: List<InputMethod>) {
		for (file in methodsDir.listFiles()!!) {
			file.delete()
		}
		for (i in inputMethods.indices) {
			val method = inputMethods[i]
			val file = File(methodsDir, i.toString() + ".json")
			storeMethod(file, method)
		}
	}

}
